core/src/main/java/io/grpc/Status.java

```diff
@@ -571,15 +562,91 @@ public String toString() {
         .toString();
   }
 
-  private static class StatusCodeMarshaller implements Metadata.AsciiMarshaller<Status> {
+  private static final class StatusCodeMarshaller implements TrustedAsciiMarshaller<Status> {
     @Override
-    public String toAsciiString(Status status) {
+    public byte[] toAsciiString(Status status) {
       return status.getCode().valueAscii();
     }
 
     @Override
-    public Status parseAsciiString(String serialized) {
-      return fromCodeValue(Integer.valueOf(serialized));
+    public Status parseAsciiString(byte[] serialized) {
+      return fromCodeValue(serialized);
+    }
+  }
+
+  private static final class StatusMessageMarshaller implements TrustedAsciiMarshaller<String> {
+
+    private static final byte[] HEX =
+        {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
+
+    @Override
+    public byte[] toAsciiString(String value) {
+      byte[] valueBytes = value.getBytes(UTF_8);
+      for (int i = 0; i < valueBytes.length; i++) {
+        byte b = valueBytes[i];
+        if (b < ' ' || b >= '~' || b == '%') {
+          return toAsciiStringSlow(valueBytes, i);
+        }
+      }
+      return valueBytes;
+    }
+
+    /**
+     * @param valueBytes the UTF-8 bytes
+     * @param ri The reader index, pointed at the first byte that needs escaping.
+     */
+    private static byte[] toAsciiStringSlow(byte[] valueBytes, int ri) {
+      byte[] escapedBytes = new byte[ri + (valueBytes.length - ri) * 3];
+      // copy over the good bytes
+      if (ri != 0) {
```

Comment:
nit: I think it's safe to assume that arraycopy will make that optimisation for you.